File: src/main/java/org/snipsniper/systray/Sniper.java
package org.snipsniper.systray;

import java.awt.*;
import java.awt.event.*;

import org.jnativehook.keyboard.NativeKeyAdapter;
import org.jnativehook.mouse.NativeMouseAdapter;
import org.snipsniper.ImageManager;
import org.snipsniper.LogManager;
import org.snipsniper.SnipSniper;
import org.snipsniper.config.ConfigHelper;

import org.jnativehook.GlobalScreen;
import org.jnativehook.keyboard.NativeKeyEvent;

import org.snipsniper.config.Config;
import org.snipsniper.capturewindow.CaptureWindow;
import org.snipsniper.configwindow.ConfigWindow;
import org.jnativehook.mouse.NativeMouseEvent;
import org.snipsniper.utils.*;
import org.snipsniper.utils.enums.LogLevel;

public class Sniper {
	public int profileID; //0 = default
	
	private CaptureWindow captureWindow;
	private ConfigWindow configWindow;
	private final Config config;
	private TrayIcon trayIcon;
	
	private final Sniper instance;

	private final NativeKeyAdapter nativeKeyAdapter;
	private final NativeMouseAdapter nativeMouseAdapter;

	public Sniper(int profileID) {
		instance = this;
		this.profileID = profileID;
		config = new Config("profile" + profileID + ".cfg", "profile_defaults.cfg");

		LogManager.log("Loading profile " + profileID, LogLevel.DEBUG);

		if(SystemTray.isSupported()) {
			Popup popup = new Popup(this);
			SystemTray tray = SystemTray.getSystemTray();

			try {
				String icon = config.getString(ConfigHelper.PROFILE.icon);
				Image image = ImageUtils.getIconDynamically(icon);

				if(image == null)
					image = ImageUtils.getDefaultIcon(profileID);
				else
					image = image.getScaledInstance(16, 16, Image.SCALE_SMOOTH);

				image.flush();
				trayIcon = new TrayIcon(image, "SnipSniper (" + getTitle() + ")");
				trayIcon.setImageAutoSize(true);

				trayIcon.addMouseListener(new MouseAdapter() {
					@Override
					public void mouseReleased(MouseEvent mouseEvent) {
						showPopup(mouseEvent);
					}

					@Override
					public void mousePressed(MouseEvent mouseEvent) {
						showPopup(mouseEvent);
					}

					private void showPopup(MouseEvent mouseEvent) {
						if (mouseEvent.isPopupTrigger())
							popup.showPopup(mouseEvent.getX(), mouseEvent.getY());
					}

					@Override
					public void mouseClicked(MouseEvent mouseEvent) {
						if (mouseEvent.getButton() == 1) {
							if (captureWindow == null && SnipSniper.isIdle()) {
								captureWindow = new CaptureWindow(instance);
								SnipSniper.setIdle(false);
							}
						}
					}
				});

				tray.add(trayIcon);
			} catch (AWTException awtException) {
				LogManager.log("There was an issue setting up the Tray Icon! Message: ", LogLevel.ERROR);
				LogManager.logStacktrace(awtException, LogLevel.ERROR);
			}
		}

		nativeKeyAdapter = new NativeKeyAdapter(){
			@Override
			public void nativeKeyPressed(NativeKeyEvent nativeKeyEvent) {
				checkNativeKey("KB", nativeKeyEvent.getKeyCode(), nativeKeyEvent.getKeyLocation());
			}
		};
		nativeMouseAdapter = new NativeMouseAdapter(){
			@Override
			public void nativeMousePressed(NativeMouseEvent nativeMouseEvent) {
				checkNativeKey("M", nativeMouseEvent.getButton(), -1);
			}
		};
		GlobalScreen.addNativeKeyListener(nativeKeyAdapter);
		GlobalScreen.addNativeMouseListener(nativeMouseAdapter);
	}

	public void kill() {
		GlobalScreen.removeNativeKeyListener(nativeKeyAdapter);
		GlobalScreen.removeNativeMouseListener(nativeMouseAdapter);
		SystemTray.getSystemTray().remove(trayIcon);
		trayIcon = null;
	}

	public void killCaptureWindow() {
		if(captureWindow != null) {
			if(SystemTray.isSupported() && getIconString().equals("none")) trayIcon.setImage(ImageManager.getImage("systray/icon" + profileID + ".png"));
			SnipSniper.setIdle(true);
			captureWindow.screenshot = null;
			captureWindow.screenshotTinted = null;
			captureWindow.isRunning = false;
			captureWindow.dispose();
			captureWindow = null;
			System.gc();
		}
	}

	public void checkNativeKey(String identifier, int pressedKey, int pressedLocation) {
		String hotkey = config.getString(ConfigHelper.PROFILE.hotkey);
		if(!hotkey.equals("NONE")) {
			if(hotkey.startsWith(identifier)) {
				int location = -1;
				if(pressedLocation != -1 && hotkey.contains("_")) {
					String[] parts =  hotkey.split("_");
					hotkey = parts[0];
					location = Integer.parseInt(parts[1]);
				}
				int key = Integer.parseInt(hotkey.replace(identifier, ""));
				if(pressedKey == key && (location == -1 || location == pressedLocation)) {
					if(captureWindow == null && SnipSniper.isIdle()) {
						captureWindow = new CaptureWindow(instance);
						SnipSniper.setIdle(false);
					}
				}
			}
		}
	}

	public String getIconString() {
		return config.getString(ConfigHelper.PROFILE.icon);
	}

	public String getTitle() {
		String title = config.getString(ConfigHelper.PROFILE.title);
		if(title.equals("none"))
			title = "Profile " + profileID;
		return title;
	}

	public Config getConfig() {
		return config;
	}

	public TrayIcon getTrayIcon() {
		return trayIcon;
	}

	public void openConfigWindow() {
		if(configWindow == null) {
			configWindow = new ConfigWindow(config, ConfigWindow.PAGE.generalPanel);
			configWindow.addCustomWindowListener(() -> configWindow = null);
		} else {
			configWindow.requestFocus();
		}
	}
}

File: src/main/kotlin/net/snipsniper/systray/Sniper.kt
package net.snipsniper.systray

import net.snipsniper.ImageManager
import net.snipsniper.NativeHookManager
import net.snipsniper.SnipSniper
import net.snipsniper.capturewindow.CaptureWindow
import net.snipsniper.config.Config
import net.snipsniper.config.ConfigHelper
import net.snipsniper.utils.*
import org.capturecoop.cclogger.CCLogLevel
import org.capturecoop.cclogger.CCLogger
import java.awt.Image
import java.awt.SystemTray
import java.awt.TrayIcon
import java.awt.event.MouseAdapter
import java.awt.event.MouseEvent

class Sniper(private val profileID: Int) {
    private var captureWindow: CaptureWindow? = null
    var config: Config = Config("profile${profileID}.cfg", "profile_defaults.cfg")
        private set
    private lateinit var trayIcon: TrayIcon
    private val instance = this

    private var isCustomIcon = false
    //TODO: Maybe make a function for opening the capture window so that we call isIdle = false at a unified place?
    //TODO: We check twice if iconString or whatever is none, we should not statically check that but maybe check with the defaults?
    init {
        CCLogger.log("Loading profile $profileID", CCLogLevel.DEBUG)
        if(SystemTray.isSupported()) {
            val popup = Popup(this)
            val tray = SystemTray.getSystemTray()
            val image = ImageUtils.getIconDynamically(config.getString(ConfigHelper.PROFILE.icon))?.scaledSmooth(16, 16).also { isCustomIcon = true } ?: getTrayIcon(profileID)
            image.flush()
            trayIcon = TrayIcon(image, "SnipSniper(${getTitle()})")
            trayIcon.isImageAutoSize = true
            trayIcon.addMouseListener(object: MouseAdapter() {
                fun showPopup(e: MouseEvent) { if(e.isPopupTrigger) popup.showPopup(e.x, e.y) }

                override fun mouseReleased(e: MouseEvent) = showPopup(e)
                override fun mousePressed(e: MouseEvent) = showPopup(e)
                override fun mouseClicked(e: MouseEvent) { if(e.button == 1) openCaptureWindow() }
            })
            tray.add(trayIcon)
        }
        //Register and listen
        if(config.getString(ConfigHelper.PROFILE.hotkey) != "NONE") {
            NativeHookManager.register(this).addListener {
                checkNativeKey(it.type.code, it.code, it.location)
            }
        }
    }

    fun kill() {
        NativeHookManager.unregister(this)
        SystemTray.getSystemTray().remove(trayIcon)
    }

    private fun checkNativeKey(identifier: String, pressedKey: Int, pressedLocation: Int) {
        var hotkey = config.getString(ConfigHelper.PROFILE.hotkey)
        if(hotkey == "NONE") return
        if(hotkey.startsWith(identifier)) {
            var location = -1
            if(pressedLocation != -1 && hotkey.contains("_")) {
                val parts = hotkey.split("_")
                hotkey = parts[0]
                location = Integer.parseInt(parts[1])
            }
            if(
                pressedKey == hotkey.replace(identifier, "").toInt() &&
                (location == -1 || location == pressedLocation)
            ) openCaptureWindow()
        }
    }

    fun getTitle(): String {
        config.getString(ConfigHelper.PROFILE.title).also {
            return if(it == config.getDefault(ConfigHelper.PROFILE.icon)) "Profile $profileID" else it
        }
    }

    private fun openCaptureWindow() {
        if(captureWindow == null && SnipSniper.isIdle) {
            if(SystemTray.isSupported() && isCustomIcon) trayIcon.image = getTrayIcon(profileID, true)
            captureWindow = CaptureWindow(instance)
            SnipSniper.isIdle = false
        } else {
            captureWindow?.requestFocus()
        }
    }

    fun killCaptureWindow() {
        if(captureWindow != null) {
            if(SystemTray.isSupported() && isCustomIcon) trayIcon.image = getTrayIcon(profileID)
            SnipSniper.isIdle = true
            captureWindow = null
            System.gc()
        }
    }

    //Note: This does not handle custom images
    private fun getTrayIcon(profileID: Int, alt: Boolean = false) = if(alt) "systray/alt_icon$profileID.png".getImage() else "systray/icon$profileID.png".getImage()

    fun alert(message: String, title: String, type: TrayIcon.MessageType) = trayIcon.displayMessage(message, title, type)

    override fun toString() = "SnipSniper Profile ($profileID)"
}